### PR TITLE
manage useResultCallback fiber within useEffect

### PR DESCRIPTION
**Pull request by tim-smart**

### CODE REVIEW

File: test/hooks/useResult.ts
Comment (datner) [inline test/hooks/useResult.ts]: thats horrendous looking lmao

File: test/hooks/useResult.ts
Comment (datner) [inline test/hooks/useResult.ts]: could you use `assert.deepEquals` (or the `expect` equivalent) or 2 assertions for this? 
Seeing 
```
expected: 1
actual: false
```
Would be rather confusing..

File: test/hooks/useResultCallback.ts
Comment (datner) [inline test/hooks/useResultCallback.ts]: same deal

File: src/hooks/useResultCallback.ts
Comment (datner) [inline src/hooks/useResultCallback.ts]: Tbh, something like this could possibly live in an `@effect/interop`, as a general abstraction for interop with many frameworks. Food for thought..

File: src/hooks/useResultCallback.ts
Comment (datner) [inline src/hooks/useResultCallback.ts]: maybe we should just leverage the [native EventTarget](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/EventTarget) for this. For 3 reasons:
1. It implements the required behavior in the native code layer
2. it [adheres to abort signals](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/removeEventListener#:~:text=Event%20listeners%20can%20also%20be%20removed%20by%20passing%20an%20AbortSignal%20to%20an%20addEventListener()%20and%20then%20later%20calling%20abort()%20on%20the%20controller%20owning%20the%20signal.), making it a possible target for more complex effect operations
3. you have much greater visibility on listerners since they show up in the listener registery. Making listener-leaks and other such bugs much easier to find and understand

it doesn't matter if you implement an event target inside the class, or have the class be an event target, I don't mind either way

**Review by datner — APPROVED** [review]
As if approving would do anything lol

File: test/hooks/useResult.ts
Comment (tim-smart) [reply test/hooks/useResult.ts]: Definitely needs improvement, there was ts errors and wanted to quickly get rid of them.

File: src/hooks/useResultCallback.ts
Comment (tim-smart) [reply src/hooks/useResultCallback.ts]: Had a look at that web api, it seems it performs worse than a simple loop. What was the api for detecting if an event is no longer being listened to?

Comment (datner) [reply src/hooks/useResultCallback.ts]: for _detecting_? not aware that's possible in general.. funny it works worse than a loop. If you find it less-suitible then the current impl is fine 

if you're talking about the 3rd point, idk take a look at the Event Listeners tab in the browser devtools, also they can be identified in the sources tab, where you can place breakpoints for listeners directly

Comment (tim-smart) [reply src/hooks/useResultCallback.ts]: We need to track the listeners so we can detect when they reach 0 (to interrupt the fiber when the consumer is unmounted), so I don't think using EventTarget is the solution here.

**Comment by tim-smart** [comment]
I'll do a bit more cleanup tomorrow, as I added the `useResult` code without removing the old version.

I think we can clean up the `TrackProperties` code too, now that both hooks have consolidated to use the store.

File: src/hooks/useResultCallback.ts
Comment (datner) [reply src/hooks/useResultCallback.ts]: You can simulate it just fine
```ts
class MyTarget extends EventTarget {
  amount = 0

  constructor() {
    super()
    super.addEventListener("cleanup", this.interruptIfRunning)
  }

  addEventListener(type: string, callback: EventListenerOrEventListenerObject | null, options?: boolean | AddEventListenerOptions | undefined): void {
      this.amount++
    super.addEventListener(type, callback, options)
  }

  removeEventListener(type: string, callback: EventListenerOrEventListenerObject | null, options?: boolean | EventListenerOptions | undefined): void {
      this.amount--
    super.removeEventListener(type,callback,options)
  }
  
  // dispatch event can use some CustomEvent extends Event type thing
  // and so on and so forth
}
```
But theres no real need to force it if we don't see a clear benefit

**Comment by datner** [comment]
for you it's tomorrow, for me it's tonight hahahah
I have some things I'm doing on my own branch, should I wait for this before I merge?

File: src/hooks/useResultCallback.ts
Comment (tim-smart) [reply src/hooks/useResultCallback.ts]: Maybe in the future - if we find other frameworks require similar workarounds.

File: src/hooks/useResultCallback.ts
Comment (tim-smart) [reply src/hooks/useResultCallback.ts]: I'm in the "lets don't force it" camp at the moment :)

**Pull request merged**
